### PR TITLE
Resolve unresolved constants in TypeExpander

### PR DESCRIPTION
**Pull request by weirdan**
This fixes vimeo/psalm#4044 by passing down `$statements_analyzer` all the way down to `getConstantForClass()`

### CODE REVIEW

File: src/Psalm/Internal/Type/TypeExpander.php [hunk 1]
```diff
@@ -33,7 +34,8 @@ public static function expandUnion(
         ?string $parent_class,
         bool $evaluate_class_constants = true,
         bool $evaluate_conditional_types = false,
-        bool $final = false
+        bool $final = false,
+        ?StatementsAnalyzer $statements_analyzer = null
```
Comment (weirdan) [inline src/Psalm/Internal/Type/TypeExpander.php:38]: Not sure if it's the best way to do it, but `expandUnion()` is just too handy.

**Comment by muglug** [comment]
I'm not sure about this particular one, mainly because I think only globally-defined constants should be referenced in a class constant.

This means constants defined in `autoload_files`, as well as the `autoload` config flag files are acceptable, but no others.

**Comment by weirdan** [comment]
Any reason why other constants won't do?

**Comment by muglug** [comment]
Any constants that are assumed to be available to _every_ class are fine IMO. It's the ones that are only available to certain classes (and thus need the `$statements_analyzer`) that trouble me.

**Comment by weirdan** [comment]
@muglug are you referring to visibility (as in `private const WHATEVER=213`) or the load order (i.e. absence of const/function autoloading in PHP)?

**Comment by muglug** [comment]
the load order – some constants are always assumed to be present if they're in `autoload_files` or defined in the `autoloader="foo.php"` file in the config

**Comment by orklah** [comment]
@weirdan seems like the original ticket was fixed. Is this still needed?

**Comment by weirdan** [comment]
@orklah I don't think the original issue is fixed, snippets there still fail in various ways. This may not be the best way to fix that though.

**Pull request closed without merging**